relays/ethereum/src/substrate_client.rs

```diff
@@ -71,301 +81,187 @@ impl Default for SubstrateSigningParams {
-	(
-		client,
-		signed_block.map(|signed_block: SignedSubstrateBlock| (id, signed_block.justification)),
-	)
 }
 
-/// Get GRANDPA authorities set at given block.
-pub async fn grandpa_authorities_set(client: Client, block: Hash) -> (Client, Result<Vec<u8>, Error>) {
-	let block = bail_on_arg_error!(to_value(block).map_err(|e| Error::RequestSerialization(e)), client);
-	call_rpc(
-		client,
-		"state_call",
-		Params::Array(vec![
-			serde_json::Value::String("GrandpaApi_grandpa_authorities".into()),
-			block,
-		]),
-		rpc_returns_bytes,
-	)
-	.await
+/// A trait for RPC calls which are used to submit Ethereum headers to a Substrate
+/// runtime. These are typically calls which use a combination of other low-level RPC
+/// calls.
+#[async_trait]
+pub trait SubmitEthereumHeaders: SubstrateRpc {
+	/// Submits Ethereum header to Substrate runtime.
+	async fn submit_ethereum_headers(
+		&self,
+		params: SubstrateSigningParams,
+		headers: Vec<QueuedEthereumHeader>,
+		sign_transactions: bool,
+	) -> Result<Vec<EthereumHeaderId>>;
+
+	/// Submits signed Ethereum header to Substrate runtime.
+	async fn submit_signed_ethereum_headers(
+		&self,
+		params: SubstrateSigningParams,
+		headers: Vec<QueuedEthereumHeader>,
+	) -> Result<Vec<EthereumHeaderId>>;
+
+	/// Submits unsigned Ethereum header to Substrate runtime.
+	async fn submit_unsigned_ethereum_headers(
+		&self,
+		headers: Vec<QueuedEthereumHeader>,
+	) -> Result<Vec<EthereumHeaderId>>;
 }
 
-/// Get Substrate block hash by its number.
-async fn block_hash_by_number(client: Client, number: Number) -> (Client, Result<Hash, Error>) {
-	let number = bail_on_arg_error!(to_value(number).map_err(|e| Error::RequestSerialization(e)), client);
-	call_rpc(
-		client,
-		"chain_getBlockHash",
-		Params::Array(vec![number]),
-		rpc_returns_value,
-	)
-	.await
-}
+#[async_trait]
+impl SubmitEthereumHeaders for SubstrateRpcClient {
+	async fn submit_ethereum_headers(
+		&self,
+		params: SubstrateSigningParams,
+		headers: Vec<QueuedEthereumHeader>,
+		sign_transactions: bool,
+	) -> Result<Vec<EthereumHeaderId>> {
+		if sign_transactions {
+			self.submit_signed_ethereum_headers(params, headers).await
+		} else {
+			self.submit_unsigned_ethereum_headers(headers).await
+		}
+	}
 
-/// Get substrate account nonce.
-async fn next_account_index(
-	client: Client,
-	account: node_primitives::AccountId,
-) -> (Client, Result<node_primitives::Index, Error>) {
-	use sp_core::crypto::Ss58Codec;
+	async fn submit_signed_ethereum_headers(
+		&self,
+		params: SubstrateSigningParams,
+		headers: Vec<QueuedEthereumHeader>,
+	) -> Result<Vec<EthereumHeaderId>> {
+		let ids = headers.iter().map(|header| header.id()).collect();
+
+		let genesis_hash = match self.genesis_hash {
+			Some(genesis_hash) => genesis_hash,
+			None => {
+				let genesis_hash = self.block_hash_by_number(Zero::zero()).await?;
+				// TODO: Fix, need &mut
+				// self.genesis_hash = Some(genesis_hash);
```

Comment:
For this, I really don't want to pass a `&mut` into this function (it bubbles up like 3 levels), so should I just use a `RefCell` or something to do this?

Reply:
Since this is just cache, it's totally fine to use `RefCell` if Send/Sync is not required for this struct.
If it is maybe use something like `once` initialization (or `lazy_static`). Even a piece of unsafe code here would be fine - we really don't care that much about consistency in the cache.

Reply:
Let's move this to constructor maybe?

Reply:
(I mean read genesis hash from `SubstrateRpcClient::new()`)

Reply:
So I wasn't able to use a `RefCell` because `Send/Sync` is required. I wasn't sure how to use `lazy_static` for this, so I didn't go that way either. I couldn't quite do what Slava suggested, since to get the genesis hash I would need an instance of `SubstrateRpcClient`, which I couldn't have in the constructor.

I added a helper instead, `SubstrateRpcClient::set_genesis_hash()`, which we call after we've created a client and used it to get the genesis hash. I do have some doubts with this approach, since I'm not sure how to handle the case where we don't get a genesis hash back (maybe the RPC connection's been interrupted).

One thing we can do is continue, but then every time we call `SubmitEthereumHeader::submit_signed_ethereum_headers()` we'll be doing an extra call to get the genesis hash, which isn't ideal. Another thing (what's in the code now) is we can just fail right there, but this also doesn't seem nice. What do you guys think?

Reply:
> I would need an instance of SubstrateRpcClient, which I couldn't have in the constructor.

What do you mean by that? It's not Java, `SubstrateRpcClient::new` is simply a builder function so it ends with you returning the instance.
I think this should definitely go to `SubstrateRpcClient::new`, but of course it is now required to be `async` - now the logic is duplicated in multiple places.
I think we should also return a `Result<>` and let the caller handle the error instead of panicking. The node might be down for instance, so the caller might choose to retry instead.

Reply:
What I mean is that in order to get the genesis block, I need to call `SubstrateRpcClient::block_hash_by_number()`, but since `new()` doesn't take `&self` I'm not able to do that